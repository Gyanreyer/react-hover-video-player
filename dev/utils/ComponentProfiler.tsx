/* eslint-disable no-console */
import React from 'react';

interface RenderTiming {
  averageRenderTime: number;
  renderCount: number;
}

interface ComponentProfilerProps {
  profilerID: string;
  children: React.ReactChild;
}

/**
 * Util component logs out render timings for whatever component it's wrapped aroud
 */
const ComponentProfiler = ({
  profilerID,
  children,
}: ComponentProfilerProps): JSX.Element => {
  // Use a ref to track render timing for this component in a persistent store
  const renderTiming = React.useRef<RenderTiming>();
  if (!renderTiming.current) {
    renderTiming.current = {
      averageRenderTime: 0,
      renderCount: 0,
    };
  }

  // Logs out helpful render timing info for performance measurements
  const onProfilerRender = React.useCallback(
    (
      id, // the "id" prop of the Profiler tree that has just committed
      phase: string, // either "mount" (if the tree just mounted) or "update" (if it re-rendered)
      actualDuration: number // time spent rendering the committed update
    ) => {
      if (phase === 'mount') {
        console.log(`${profilerID} | MOUNT: ${actualDuration}ms`);
      } else {
        renderTiming.current.renderCount += 1;
        renderTiming.current.averageRenderTime +=
          (actualDuration - renderTiming.current.averageRenderTime) /
          renderTiming.current.renderCount;
        console.log(
          `${profilerID} | UPDATE: ${actualDuration}ms | New average: ${renderTiming.current.averageRenderTime}ms`
        );
      }
    },
    [profilerID]
  );
  return (
    <React.Profiler id={profilerID} onRender={onProfilerRender}>
      {children}
    </React.Profiler>
  );
};

export default ComponentProfiler;
