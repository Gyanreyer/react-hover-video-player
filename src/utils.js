/**
 * @typedef   {object}  VideoSource
 * @property  {string}  src - The src URL string to use for a video player source
 * @property  {string}  type - The media type of the video, ie 'video/mp4'
 */

/**
 * @function  formatVideoSrc
 *
 * Takes a videoSrc value and formats it as an array of VideoSource objects which can be used to render
 * <source> elements for the video
 *
 * @param {!(string|string[]|VideoSource|VideoSource[])}  videoSrc - Source(s) to use for the video player. Accepts 3 different formats:
 *                                                                   - **String**: the URL string to use as the video player's src
 *                                                                   - **Object**: an object with attributes:
 *                                                                     - src: The src URL string to use for a video player source
 *                                                                     - type: The media type of the video source, ie 'video/mp4'
 *                                                                   - **Array**: if you would like to provide multiple sources, you can provide an array of URL strings and/or objects with the shape described above
 *
 * @returns {VideoSource[]} Array of formatted VideoSource objects which can be used to render <source> elements for the video
 */
export function formatVideoSrc(videoSrc) {
  if (videoSrc == null) {
    // A videoSrc value is required in order to make the video player work
    console.error(
      "Error: 'videoSrc' prop is required for HoverVideoPlayer component"
    );

    return [];
  }

  return (
    // Make sure we can treat the videoSrc value as an array
    []
      .concat(videoSrc)
      // Parse our video source values into an array of VideoSource objects that can be used to render sources for the video
      .reduce((sourceArray, source) => {
        if (typeof source === 'string') {
          // If the source is a string, it's an src URL so format it into a VideoSource object and add it to the array
          sourceArray.push({ src: source });
        } else if (source && source.src) {
          // If the source is an object with an src, just add it to the array
          sourceArray.push({ src: source.src, type: source.type });
        } else {
          // Log an error if one of the videoSrc values is invalid
          console.error(
            "Error: invalid value provided to HoverVideoPlayer prop 'videoSrc':",
            source
          );
        }

        return sourceArray;
      }, [])
  );
}

/**
 * @typedef   {object}  VideoCaptionsTrack
 * @property  {string}  src - The src URL string for the captions track file
 * @property  {string}  srcLang - The language code for the language that these captions are in
 * @property  {string}  label - The title of the captions track
 * @property  {bool}    default - Whether this track should be used by default if the user's preferences don't match an available srcLang
 */

/**
 * @function formatVideoCaptions
 *
 * Takes a videoCaptions value and formats it as an array of VideoCaptionsTrack objects which can be used to render
 * <track> elements for the video
 *
 * @param {!(string|string[]|VideoCaptionsTrack|VideoCaptionsTrack[])} videoCaptions - Captions track(s) to use for the video player for accessibility. Accepts 3 different formats:
 *                                                                                     - **String**: the URL string to use as the captions track's src
 *                                                                                     - **Object**: an object with attributes:
 *                                                                                       - src: The src URL string for the captions track file
 *                                                                                       - srcLang: The language code for the language that these captions are in (ie, 'en', 'es', 'fr')
 *                                                                                       - label: The title of the captions track
 *                                                                                       - default: Whether this track should be used by default if the user's preferences don't match an available srcLang
 *                                                                                     - **Array**: if you would like to provide multiple caption tracks, you can provide an array of objects with the shape described above
 *
 * @returns {VideoCaptionsTrack[]}  Array of formatted VideoCaptionsTrack objects which can be used to render <track> elements for the video
 */
export function formatVideoCaptions(videoCaptions) {
  // If no captions were provided, return an empty array
  if (videoCaptions == null) return [];

  return (
    // Make sure we can treat the videoSrc value as an array
    []
      .concat(videoCaptions)
      // Parse our video captions values into an array of VideoCaptionsTrack
      // objects that can be used to render caption tracks for the video
      .reduce((captionsArray, captions) => {
        if (captions && captions.src) {
          captionsArray.push({
            src: captions.src,
            srcLang: captions.srcLang,
            label: captions.label,
            default: Boolean(captions.default),
          });
        } else {
          // Log an error if one of the videoCaptions values is invalid
          console.error(
            "Error: invalid value provided to HoverVideoPlayer prop 'videoCaptions'",
            captions
          );
        }

        return captionsArray;
      }, [])
  );
}
