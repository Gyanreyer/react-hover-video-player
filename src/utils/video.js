/* eslint-disable no-param-reassign */
// Enumerates states that the video can be in
export const VIDEO_STATE = {
  paused: 'paused',
  loading: 'loading',
  playing: 'playing',
};

/**
 * @function getVideoState
 *
 * Takes a video element and returns its current playing state
 *
 * @param {node} videoElement
 */
export function getVideoState(videoElement) {
  if (videoElement.paused || videoElement.ended) {
    return VIDEO_STATE.paused;
  }

  // If the video isn't paused but its readyState indicates it isn't loaded enough
  // to play yet, it is loading
  if (videoElement.readyState < 3) {
    return VIDEO_STATE.loading;
  }

  // If the video isn't paused and its ready state indicates it's loaded enough to play,
  // assume it's playing
  return VIDEO_STATE.playing;
}

/**
 * @function playVideo
 *
 * Takes a video element and attempts to play it.
 *
 * @param {node}  videoElement - The video element to start playing
 *
 * @returns {Promise}   A Promise which will resolve when the video starts playing or reject if an error occurs
 */
export function playVideo(videoElement) {
  // Start playing the video and hang onto the play promise it returns
  let playPromise = videoElement.play();

  if (!playPromise || !playPromise.then) {
    // If videoElement.play() didn't return a promise, we'll manually create one
    // ourselves which mimics the same behavior
    playPromise = new Promise((resolve, reject) => {
      // Declaring onVideoPlaybackFailed up here so we can refer to it and remove its event listener
      // if the video successfully starts playing
      let onVideoPlaybackFailed;

      // Set up event listener to resolve the promise when the video player starts playing
      const onVideoPlaybackStarted = () => {
        // Remove the event listeners we added as cleanup now that the play attempt has succeeded
        videoElement.removeEventListener('playing', onVideoPlaybackStarted);
        videoElement.removeEventListener('error', onVideoPlaybackFailed);

        // Resolve because we successfully started playing!
        resolve();
      };
      videoElement.addEventListener('playing', onVideoPlaybackStarted, {
        once: true,
      });

      // Set up event listener to reject the promise when the video player encounters an error
      onVideoPlaybackFailed = (event) => {
        // Remove the event listeners we added as cleanup now that the play attempt has failed
        videoElement.removeEventListener('error', onVideoPlaybackFailed);
        videoElement.removeEventListener('playing', onVideoPlaybackStarted);

        // Reject with the error that was thrown
        reject(event.error);
      };
      videoElement.addEventListener('error', onVideoPlaybackFailed);
    });
  }

  return playPromise;
}

/**
 * @function  pauseVideo
 *
 * Pauses the video and resets it to the beginning if necessary
 *
 * @param {node}  videoElement - The video element to pause playback for
 * @param {bool}  shouldRestart - Whether the videp should be reset to the beginning after pausing
 */
export function pauseVideo(videoElement, shouldRestart) {
  videoElement.pause();

  if (shouldRestart) {
    // If we should restart the video, reset its time to the beginning
    videoElement.currentTime = 0;
  }
}
