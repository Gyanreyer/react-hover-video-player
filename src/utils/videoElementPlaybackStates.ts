/**
 * Takes a video element and returns whether it is currently paused, meaning it is not
 * playing or attempting to play.
 *
 * @param {HTMLVideoElement} videoElement
 */
export function isVideoElementPaused(
  videoElement: HTMLVideoElement | null
): boolean {
  return videoElement ? videoElement.paused || videoElement.ended : true;
}
