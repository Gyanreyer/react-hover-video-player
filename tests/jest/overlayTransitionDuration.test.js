import React from 'react';
import { fireEvent } from '@testing-library/react';

import { renderHoverVideoPlayer, advanceVideoTime } from './utils';

describe('overlayTransitionDuration', () => {
  afterEach(() => {
    expect(console.error).not.toHaveBeenCalled();
  });

  test('Stop attempts take the amount of time set by overlayTransitionDuration prop if a pausedOverlay is provided', () => {
    const {
      videoElement,
      playerContainer,
      pausedOverlayWrapper,
    } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      overlayTransitionDuration: 900,
    });

    expect(videoElement).toBePaused();
    // Paused overlay should be visible and have correct initial styles
    expect(pausedOverlayWrapper).toHaveStyle({
      opacity: 1,
      pointerEvents: 'auto',
      // Transition duration should match the overlayTransitionDuration prop value
      transition: 'opacity 900ms',
    });

    // Hover to start playing the video
    fireEvent.mouseEnter(playerContainer);

    //  Advance time for the video to load and start playing
    advanceVideoTime(300);
    expect(videoElement).toBePlaying();
    expect(pausedOverlayWrapper).not.toBeVisible();

    // Mouse out to stop the video
    fireEvent.mouseLeave(playerContainer);

    // The pause overlay should be visible/fading back in
    expect(pausedOverlayWrapper).toBeVisible();

    // We don't want to pause the video until the pause overlay's 900ms CSS transition has completed
    expect(videoElement.pause).not.toHaveBeenCalled();
    expect(videoElement).toBePlaying();

    // Advance timer just up to before the pause timeout should complete
    advanceVideoTime(899);

    // The pause timeout should still be in progress since we're just 1ms short of the transition duration
    expect(videoElement.pause).not.toHaveBeenCalled();
    expect(videoElement).toBePlaying();

    // Advance time by 1ms so the pause timeout can run
    advanceVideoTime(300);

    // The stop attempt should be completed now that the full fade duration is complete
    expect(videoElement.pause).toHaveBeenCalledTimes(1);
    expect(videoElement).toBePaused();
  });

  test('Stop attempts stop immediately if a pausedOverlay is not provided', () => {
    const { videoElement, playerContainer } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      overlayTransitionDuration: 900,
    });

    // Get the video playing
    expect(videoElement).toBePaused();
    fireEvent.mouseEnter(playerContainer);
    advanceVideoTime(300);
    expect(videoElement).toBePlaying();

    // Stop the video
    fireEvent.mouseLeave(playerContainer);

    // The video should have been paused immediately since we don't have a pause overlay
    expect(videoElement.pause).toHaveBeenCalledTimes(1);
    expect(videoElement).toBePaused();
  });
});
