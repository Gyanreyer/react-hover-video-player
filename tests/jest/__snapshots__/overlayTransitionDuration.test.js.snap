// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overlayTransitionDuration Stop attempts stop immediately if a pausedOverlay is not provided 1`] = `
<div>
  <div
    data-testid="hover-video-player-container"
    style="position: relative;"
  >
    <video
      crossorigin="anonymous"
      data-testid="video-element"
      loop=""
      playsinline=""
      style="display: block; width: 100%; object-fit: cover;"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`overlayTransitionDuration Stop attempts take the amount of time set by overlayTransitionDuration prop if a pausedOverlay is provided 1`] = `
<div>
  <div
    data-testid="hover-video-player-container"
    style="position: relative;"
  >
    <div
      data-testid="paused-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 1; opacity: 1; transition: opacity 900ms; pointer-events: auto;"
    >
      <div />
    </div>
    <video
      crossorigin="anonymous"
      data-testid="video-element"
      loop=""
      playsinline=""
      style="display: block; width: 100%; object-fit: cover;"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;
